virtool/ws/server.py

```diff
@@ -78,6 +80,37 @@ def remove_connection(self, connection: WSConnection):
         except ValueError:
             pass
 
+    async def handle_redis_ws_messages(self):
+        channel = self._redis.subscribe("close_websocket_connections")
+
+        while await channel.wait_message():
+            session_id = await channel.get(encoding="utf-8")
+            await self.close_connections_with_session_id(session_id)
+
+    async def close_connections_with_session_id(self, session_id):
+        """
+        Closes connections associated with a session_id
+        """
+        for connection in self._connections:
+            if connection.session_id == session_id:
+                await connection.close(1000)
+
+    async def periodically_close_expired_websocket_connections(self):
+        """
+        Periodically closes connections with expired sessions.
+        """
+        session_data = SessionData(self._redis)
+
+        while True:
+            for connection in self._connections:
+                try:
+                    session = await session_data.get_anonymous(connection.session_id)
+                    if session.authentication is None:
+                        await connection.close(1001)
```

Comment:
Couple of points:
 1. We can assume that if a connection exists then it was made using an authenticated session
 2. `get_anonymous` should only have ever returned anonymous (non-authenticated) sessions, this has been fixed on latest

Bring your branch up to date and re-implement around the fixed sessions datalayer. You may need to add a new dataLayer function to check if a `session_id` is associated with an authenticated session without having a session token.